src/LockstakeClipper.sol

```
@@ -354,44 +367,44 @@ contract LockstakeClipper {
                 // Total debt will be paid
-                owe = tab;                  // owe' <= owe
+                owe = sale.tab;                  // owe' <= owe
                 // Adjust slice
-                slice = owe / price;        // slice' = owe' / price <= owe / price == slice <= lot
-            } else if (owe < tab && slice < lot) {
+                slice = owe / price;             // slice' = owe' / price <= owe / price == slice <= lot
+            } else if (owe < sale.tab && slice < sale.lot) {
                 // If slice == lot => auction completed => dust doesn't matter
                 uint256 _chost = chost;
-                if (tab - owe < _chost) {    // safe as owe < tab
+                if (sale.tab - owe < _chost) {   // safe as owe < tab
                     // If tab <= chost, buyers have to take the entire lot.
-                    require(tab > _chost, "LockstakeClipper/no-partial-purchase");
+                    require(sale.tab > _chost, "LockstakeClipper/no-partial-purchase");
                     // Adjust amount to pay
-                    owe = tab - _chost;      // owe' <= owe
+                    owe = sale.tab - _chost;     // owe' <= owe
                     // Adjust slice
-                    slice = owe / price;     // slice' = owe' / price < owe / price == slice < lot
+                    slice = owe / price;         // slice' = owe' / price < owe / price == slice < lot
                 }
             }
 
             // Calculate remaining tab after operation
-            tab = tab - owe;  // safe since owe <= tab
+            sale.tab = sale.tab - owe;  // safe since owe <= tab
             // Calculate remaining lot after operation
-            lot = lot - slice;
+            sale.lot = sale.lot - slice;
 
             // Send collateral to who
             vat.slip(ilk, address(this), -int256(slice));
-            engine.onTake(usr, who, slice);
+            engine.onTake(sale.usr, who, slice);
```

As was brought up at some point, do we want to add the cuttee to the list below of addresses that we can not do `clipperCall()` to?
If not we should at lest add a comment.

---

Yes I still need to do that

---

So we have two ways to do it:

1-
```
data.length > 0 &&
who != address(0) &&
who != address(vat) &&
who != address(dog_) &&
who != address(engine) &&
who != cuttee
```

or

2-
```
data.length > 0 &&
who != address(vat) &&
who != address(dog_) &&
who != address(engine) &&
(cuttee == address(0) || who != cuttee)
```

I prefer option 1 as looks simpler and nicer, although option 2 doesn't change the actual behaviour (if you pass data and who is address(0) reverts). Looking for opinions.

---

Honestly I don't see a big difference between the two options. Maybe you don't want to change behavior at this point but IMO it's actually cleaner to revert when a callee is disallowed rather than silently skipping the call:

```solidity
if (data.length > 0) {
    if (who == address(vat) || who == address(dog_) || who == address(engine) || who == cuttee) {
          revert DisallowedCallee();
    }
    ClipperCallee(who).clipperCall(msg.sender, owe, slice, data);
}
```

If you care enough about not changing behavior not to go with that suggestion, then probably you also want to go with the 2nd option above to preserve as much behavior as possible.

---

I would avoid changing behavior vs the other deployed clippers as much as we can.
Would prefer option 2, so we keep the implicit restriction that if someone passes `who == adress(0)` they also pass empty `data`.

---

Option 2 implemented.

---

would it make sense to do `(who != cuttee || cuttee == address(0))` instead of `(cuttee == address(0) || who != cuttee)` ?

This would cost a bit less gas sine the left expression is evaluated first and we expect it to be true, so it would short circuit and not do the right check.

Could also change `cuttee == address(0)` to `who == address(0)` to avoid the (warm) storage read, but if we change the order as suggested above then I don't expect we reach that evaluation anyway on the normal path.

---

Changed